### DISCUSSION TITLE
Memory leaks and RequestContext

### DISCUSSION THREAD

**Discussion by roderik** [mikro-orm/mikro-orm#2063] [Q&A] [answered]
I’ve been hunting a memory leak in my nest application for a few days, and I think I’m fairly sure that I do not understand how the RequestContext is supposed to work. 

Given the following cron job, I thought it should (since the decorator is there) create a new context, use the em from this clean context, and then this cron is done, garbage collect. (maybe with some delay)

What I’m seeing is my heap space jumping 3-400MB, but never dropping (maybe a fraction).

I see the same for my requests, every request i can check the RequestContext.currentRequestcontext.id and get a new id, but a call with a simple resolver->service->repository.findAndCount()  (returns 6 complex items) loses 1-2MB.

```
  constructor(
    @Inject(forwardRef(() => ApplicationEntityService))
    protected applicationService: ApplicationEntityService,
    @InjectRepository(Usage)
    private readonly usageRepository: UsageRepository,
  ) {}

  @Cron(CronExpression.EVERY_HOUR)
  @UseRequestContext()
  async ensureStripeSubscriptionItemCron() {
    await this.applicationService.list() // does this.applicationRepository.findAll() in the service
    await this.usageRepository.findAll()
  }
```

I've removed about everything from my app that could interfere, and it still happens.

What am I missing here?

**Comment by roderik** [comment]
To give some idea, this is just the cronjob every hour

![CleanShot 2021-07-25 at 21 16 28](https://user-images.githubusercontent.com/16780/126910718-243e9d32-35ca-4467-94bc-f2f20f508746.png)

**Comment by roderik** [comment]
Trying some snapshots, i see 7 em's are retained with 61k of the usage entities and 61k of other wrapped entities, they are all retained by the mongo driver if i read this correct.

![CleanShot 2021-07-25 at 22 30 15](https://user-images.githubusercontent.com/16780/126912643-2e50b8ca-ae24-4f0e-bc56-0ab277e92f5a.png)
![CleanShot 2021-07-25 at 22 31 53](https://user-images.githubusercontent.com/16780/126912662-1c07e60a-53d9-4e26-b2f3-3ef2e86c872a.png)
![CleanShot 2021-07-25 at 22 32 09](https://user-images.githubusercontent.com/16780/126912676-b07d792b-fcb6-4504-a9e1-1c8afb40ee0f.png)

**Comment by jgburet** [comment]
You need to not use the first entity manager the orm give you once connected to your database, but instead fork it.

For example, each time you receive a new http request, in your Express/Koa/whatever middlewares chain, make sure to pass to the code downstream a child of the root entity manager. You can get one by doing `em.fork()`.
Of course, this works the same each time you have process workers getting a task from your job queue, or each time you run a CLI command.

**Reply by roderik** [reply]
I can confirm i get new RequestContext id's every time. In the screenshots above you will also see the domain mentioned in the retainers.

Not an expert in memory traces, but it seems the mongo driver is keeping a reference.

**Comment by roderik** [comment]
I was hoping mongo would GC itself at some point, but no (drops are restarts due to k8s memory limits)

![CleanShot 2021-07-26 at 08 24 25](https://user-images.githubusercontent.com/16780/126942571-063bcc42-f7ba-4dd0-9bbb-912397a63957.png)

**Comment by roderik** [comment]
No improvement from downgrading a few versions of mikro, nor switching to AsyncLocalStorage

**Reply by roderik** [reply]
AsyncLocalStorage is not for the @UseRequestContext annotation, so duh....

**Comment by roderik** [comment, marked as answer]
In the end:

- not loading 62k items every 5 seconds helps a lot
- concurrently, so not really reproducible, but removing Sentry Tracing (uses a request context and mongo instrumentation) also seem to have helped a lot
